### PR TITLE
[HOLD] Don't include null in collection callbacks

### PR DESCRIPTION
**Pull request by roryabraham**
### Details
Coming from https://github.com/Expensify/App/issues/25481#issuecomment-1687321203, we are including null in `Onyx.connect` subscriber callbacks if it's a collection key w/ `waitForCollectionCallback: true` 🐛 

This PR adds a test to cover this edge case and then prevents it such that the full collection w/o null is given to the subscriber.

### Related Issues
Fixes https://github.com/Expensify/react-native-onyx/issues/310

### Automated Tests
Added automated test to cover this.

### Linked PRs
Not created yet

### CODE REVIEW

File: lib/Onyx.js [hunk 1]
```diff
@@ -509,7 +509,9 @@ function keyChanged(key, data, canUpdateSubscriber) {
         if (_.isFunction(subscriber.callback)) {
             if (isCollectionKey(subscriber.key) && subscriber.waitForCollectionCallback) {
                 const cachedCollection = getCachedCollection(subscriber.key);
-                cachedCollection[key] = data;
+                if (data !== null) {
+                    cachedCollection[key] = data;
+                }
```
Comment (Pujan92) [inline lib/Onyx.js:514]: For data `null`, don't we need to remove that key from the `cachedCollection` as we are not updating with any other value?

Comment (marcaaron) [reply lib/Onyx.js:514]: I agree. Gonna ask the question another way:

Why do we need to set `data` to the `key` for the `cachedCollection`? Shouldn't the cache _already_ be correct by this point? If we are adding a key it should be there. If we are removing a key it should be there.

File: tests/unit/onyxTest.js [hunk 2]
```diff
@@ -53,6 +53,22 @@ describe('Onyx', () => {
             expect(keys.includes(ONYX_KEYS.OTHER_TEST)).toBe(false);
         }));
 
+    it('should not provide null values to subscriber callback when set is called with null value', () => {
```
Comment (marcaaron) [inline tests/unit/onyxTest.js:56]:
```suggestion
    it('should not provide null values to subscriber callback using waitForCollectionCallback when set is called with null value', () => {
```

File: tests/unit/onyxTest.js [hunk 3]
```diff
@@ -53,6 +53,22 @@ describe('Onyx', () => {
...
+        const subscriberCallback = jest.fn();
+        Onyx.connect({
+            key: ONYX_KEYS.COLLECTION.TEST_CONNECT_COLLECTION,
+            waitForCollectionCallback: true,
+            callback: subscriberCallback,
+        });
+        const itemID = 'abcd';
+        const testKey = `${ONYX_KEYS.COLLECTION.TEST_CONNECT_COLLECTION}${itemID}`;
+        return Onyx.set(testKey, 42)
```
Comment (marcaaron) [inline tests/unit/onyxTest.js:65]: Should we add tests for the `merge()` and `mergeCollection()` cases?

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: tests/unit/onyxTest.js (hunk 3 above)
Comment (roryabraham) [reply tests/unit/onyxTest.js]: Done! This exercise uncovered that `merge` and `mergeCollection` weren't really behaving how I would expect

File: tests/unit/onyxTest.js (hunk 2 above)
Comment (roryabraham) [reply tests/unit/onyxTest.js]: done

File: lib/Onyx.js (hunk 1 above)
Comment (roryabraham) [reply lib/Onyx.js:515]: Added assertions to ensure that keys that are being deleted due to being null are dropped from both cache and storage.

File: lib/Onyx.js [hunk 4]
```diff
@@ -634,7 +636,7 @@ function sendDataToConnection(mapping, val, matchedKey) {
         return;
     }
 
-    if (_.isFunction(mapping.callback)) {
+    if (_.isFunction(mapping.callback) && (isCollectionKey(mapping.key) || matchedKey)) {
```
Comment (roryabraham) [inline lib/Onyx.js:639]: This change prevents extra calls to `Onyx.connect` subscribers with value `null` and key `undefined`. `withOnyx` waits for all connection callbacks to have been called once before rendering, so that's why we added this type of callback. However, I don't think that this type of callback is helpful to other library code using `Onyx.connect` directly, so I prevented that here. `withOnyx` callbacks are handled in the block above which terminates with an early return.

File: tests/unit/onyxTest.js [hunk 5]
```diff
@@ -671,14 +802,11 @@ describe('Onyx', () => {
                 return waitForPromisesToResolve();
             })
             .then(() => {
-                // Then we expect the callback to have called twice, once for the initial connect call + once for the collection update
-                expect(mockCallback).toHaveBeenCalledTimes(2);
-
-                // AND the value for the first call should be null since the collection was not initialized at that point
-                expect(mockCallback).toHaveBeenNthCalledWith(1, null, undefined);
```
Comment (roryabraham) [inline tests/unit/onyxTest.js:678]: This is an instance of the `null, undefined` callback mentioned above that I removed.

File: .eslintrc.js [hunk 6]
```diff
@@ -15,4 +15,13 @@ module.exports = {
         },
     },
     ignorePatterns: 'dist',
+    overrides: [
```
Comment (roryabraham) [inline .eslintrc.js:18]: Discussion:

- https://expensify.slack.com/archives/C01GTK53T8Q/p1692889359641599
- https://expensify.slack.com/archives/C03TQ48KC/p1692889446895489

Also, this same change is also present in E/App: https://github.com/Expensify/App/blob/f74b3325b3622f8afdee44ce8de1a40b8895c1bc/.eslintrc.js#L166-L173

File: lib/Onyx.js (hunk 4 above)
Comment (marcaaron) [reply lib/Onyx.js:639]: Can you clarify which call to `sendDataToConnection()` we are trying to prevent and why? There are only a few of them - so I worry that we are adding this conditional exception in the wrong place. And maybe don't really need this change at all.

File: lib/Onyx.js [hunk 7]
```diff
@@ -1019,6 +1021,10 @@ function multiSet(data) {
  * @returns {*}
  */
 function applyMerge(existingValue, changes) {
+    if (_.some(changes, change => _.isNull(change))) {
+        return null;
+    }
```
Comment (marcaaron) [inline lib/Onyx.js:1026]: This seems wrong to me. If _any_ of the changes in the merge queue array are `null` then we `null` out the resulting value?

File: lib/Onyx.js [hunk 8]
```diff
@@ -1080,6 +1086,11 @@ function merge(key, changes) {
                 // After that we merge the batched changes with the existing value
                 const modifiedData = removeNullObjectValues(applyMerge(existingValue, [batchedChanges]));
 
+                // If the data is null, remove it from cache and store
+                if (_.isNull(modifiedData)) {
+                    return remove(key);
+                }
```
Comment (marcaaron) [inline lib/Onyx.js:1092]: I think this can happen sooner. If the `batchedChanges` are `null` then we don't need to concern ourselves with the `existingValue` because we are deleting everything anyway.

File: lib/Onyx.js [hunk 9]
```diff
@@ -1248,19 +1259,34 @@ function mergeCollection(collectionKey, collection) {
...
-                .value();
+            const existingKeysToKeep = [];
+            const keysToRemove = [];
+            const collectionWithoutKeysToRemove = collection;
+            const keyValuePairsForExistingCollection = [];
+            const keyValuePairsForNewCollection = [];
+            _.each(collection, (value, key) => {
+                if (value === null) {
+                    keysToRemove.push(key);
+                    delete collectionWithoutKeysToRemove[key];
```
Comment (marcaaron) [inline lib/Onyx.js:1270]: Rather than delete the keys it might be safer to add the ones we are keeping to a new object and avoid mutating the object passed to `mergeCollection()` (I think maybe we are mutating it).

File: lib/Onyx.js [hunk 10]
```diff
@@ -1248,19 +1259,34 @@ function mergeCollection(collectionKey, collection) {
...
+                }
+
+                keyValuePairsForNewCollection.push([key, removeNullObjectValues(value)]);
+            });
 
             const promises = [];
 
+            if (keysToRemove.length > 0) {
+                _.each(keysToRemove, key => cache.drop(key));
+                promises.push(Storage.removeItems(keysToRemove));
```
Comment (marcaaron) [inline lib/Onyx.js:1287]: Why not use `Onyx.remove()` here like we did in `Onyx.merge()`?

File: lib/Onyx.js [hunk 11]
```diff
@@ -1248,19 +1259,34 @@ function mergeCollection(collectionKey, collection) {
...
+                    return;
+                }
 
-            const existingKeyCollection = _.pick(collection, existingKeys);
-            const newCollection = _.pick(collection, newKeys);
-            const keyValuePairsForExistingCollection = prepareKeyValuePairsForStorage(existingKeyCollection);
-            const keyValuePairsForNewCollection = prepareKeyValuePairsForStorage(newCollection);
+                if (persistedKeys.includes(key)) {
+                    existingKeysToKeep.push(key);
+                    keyValuePairsForExistingCollection.push([key, removeNullObjectValues(value)]);
```
Comment (marcaaron) [inline lib/Onyx.js:1276]: Hmm using `removeNullObjectValues()` seems like a bug here. Wouldn't that mean that when we merge an object and properties that are set to `null` (presumably to remove those top-level values) they will keep existing?

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: lib/Onyx.js (hunk 4 above)
Comment (roryabraham) [reply lib/Onyx.js:639]: Yes, if you do this:

```js
const myCallback = () => {};
Onyx.connect({
   key: 'someKey',
   callback: myCallback,
});
```

then `myCallback` will _always_ first be called with `null, undefined` for the value and key. I think this is counter-intuitive and unhelpful in most cases.

The reason we do that is because `withOnyx` waits for each `Onyx.connect` callback to be called at least once before rendering the child component. So even if the data is missing, we call it with `null, undefined`.

My thinking is that by doing this we're exposing an internal implementation detail of `withOnyx` unnecessarily, and in the general case `Onyx.connect` should only execute the callback when there's some data in the store with the given key.

File: lib/Onyx.js [hunk 12]
```diff
@@ -488,7 +488,7 @@ function keysChanged(collectionKey, partialCollection) {
  * @param {*} data
  * @param {Function} [canUpdateSubscriber] only subscribers that pass this truth test will be updated
  */
-function keyChanged(key, data, canUpdateSubscriber) {
+function keyChanged(key, data, canUpdateSubscriber = () => true) {
```
[the remark this answers is not in the document]
Comment (roryabraham) [reply lib/Onyx.js:491]: I agree this wasn't really needed for this PR, but I disagree that it was more obvious before...

Before:
- param can be function or undefined
- no default for an optional param. in order to understand the default behavior, you need to read the function and see how it's implemented

After:
- param is always a function
- obvious default for an optional param. can be understood just by looking at the function definition

File: lib/Onyx.js (hunk 10 above)
Comment (roryabraham) [reply lib/Onyx.js:1287]: Just because we're bulk-removing items with `Storage.removeItems` instead of multiple calls to `Storage.removeItem`

File: lib/Onyx.js (hunk 7 above)
Comment (roryabraham) [reply lib/Onyx.js:1026]: I am not 100% sure about this one, but I _think_ it makes sense? i.e: if I'm trying to merge all these changes into an empty object:

```js
[
    {a: 1},
    {b: 2},
    null,
    {c: 3},
]
```

then what should the result be? This code makes it `null`, but maybe it should actually be applied in order such such that the final result is `{c: 3}`?

File: lib/Onyx.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (dangrous) [reply lib/Onyx.js:491]: I'd lean towards this new version as well, as I agree it's more understandable at a glance!

File: lib/Onyx.js [hunk 13]
```diff
@@ -1020,16 +1022,20 @@ function multiSet(data) {
  */
 function applyMerge(existingValue, changes) {
     const lastChange = _.last(changes);
+    if (_.isNull(lastChange)) {
```
Comment (dangrous) [inline lib/Onyx.js:1025]: Can you combine this with the if() on new line 1029? Has the same result

File: lib/Onyx.js [hunk 14]
```diff
@@ -1020,16 +1022,20 @@ function multiSet(data) {
...
+        return lastChange;
+    }
 
     if (_.isArray(existingValue) || _.isArray(lastChange)) {
         return lastChange;
     }
 
-    if (_.isObject(existingValue) || _.every(changes, _.isObject)) {
+    const changesToMake = _.filter(changes, change => Boolean(change));
```
Comment (dangrous) [inline lib/Onyx.js:1033]: NAB - might be good to leave a comment here explaining the Boolean thing (i.e. I *think* I understand this - that it's filtering out any null/undefineds? - but I'm not 100% sure so others might not be either)

File: lib/Onyx.js (hunk 9 above)
Comment (dangrous) [reply lib/Onyx.js:1270]: This makes sense to me

File: lib/Onyx.js (hunk 11 above)
Comment (dangrous) [reply lib/Onyx.js:1276]: Different comment, same line - can we just have `existingKeysToKeep.push(key);` in this if, and then use the `keyValuePairsForExistingCollection.push` on line 1280, instead of doubling it here?

Assuming this is still what we want to do, after @marcaaron's comment

**Review by dangrous — CHANGES_REQUESTED** [review]
some additional comments outside of the main discussions

File: lib/Onyx.js (hunk 4 above)
Comment (marcaaron) [reply lib/Onyx.js:639]: Sounds like the main premise in your argument is that "no Onyx.connect() subscribers need this". 

**thought:** passing `null` also provides a way for an Onyx.connect() subscriber to see that it initialized _without a value from storage_. 

Is it helpful to know this? Do we have existing code that depends on this information? I think it's safer to assume we do than don't, but also don't want to audit the 446 usages 😄

File: lib/Onyx.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply lib/Onyx.js:491]: it's not a hill I will die on. I will simply change it back to what _I prefer_ in my next Onyx PR :trollface:

File: lib/Onyx.js (hunk 7 above)
Comment (marcaaron) [reply lib/Onyx.js]: >then what should the result be?

It should be the last value in the array

Comment (marcaaron) [reply lib/Onyx.js]: >If null is anywhere else in the array, only the last item in the array will be applied.

Not anywhere else in the array. Specifically, it should clear out the items that came before it. The rest should merge normally.

Comment (marcaaron) [reply lib/Onyx.js]: >If null is anywhere else in the array, it is ignored.

I think this logic is wrong. It should be clearing the previous values called in the same tick of the event loop. 

If we do this:

```js
Onyx.merge('test', {a: 1});
Onyx.merge('test', null);
Onyx.merge('test', {b: 2});
Onyx.merge('test', {c: 3});
```
Then the expected result would be `{b:2, c: 3}`?

**Comment by roryabraham** [comment]
Thanks for the reviews here everyone. While I think this is a valuable improvement, as far as I know there are no active bugs associated with the current behavior. So I'm going to reprioritize a bit and probably not pick this PR back up until next week. Thanks for understanding 🙇🏼

**Comment by marcaaron** [comment]
@roryabraham still focusing on this one? or should we close it and reopen later?

**Comment by roryabraham** [comment]
Haven't focused on it for a bit, but I think it could be a good one to pick back up. Maybe just a holistic look at how deleting stuff from Onyx works today and how we expect everything to work would be a good start

**Comment by tgolen** [comment]
I'm placing this on HOLD right now in order for QA to be done on the latest version of Onyx without more un-QAed PRs from being merged.

**Comment by roryabraham** [comment]
Closing this as not planned for now. I believe there are separate efforts to make merging / setting null more consistent
